Documentation for OQL (#8431)
mendix/docs: content/en/docs/refguide/modeling/domain-model/oql/oql-expression-syntax.md

* AND
* OR

### NULL Handling

If one of the `expression` in a binary operation or the unary `expression` have a `NULL` value, then the return type will also be NULL.

This does not apply to the `=` and `!=` operators. Handling of `NULL` in [other operators](#other-operators) is detailed in the specific operator subsections.

## String Coercion

In some databases, using `STRING` type variables in place of numeric, `DATETIME` or `BOOLEAN` values in operators and functions that explicitly require those types, causes the database to perform an implicit conversion. A common example would be the use of a `STRING` representation of a `DATETIME` variable inside a `DATEPART` function. Mendix recommends that you always [cast](#cast) strings to the exact type the operator or functions.

## Functions

These are the currently supported functions:

* CAST
* COALESCE
* DATEDIFF
* DATEPART
* LENGTH
* LOWER
* RANGEBEGIN
* RANGEEND
* REPLACE
* ROUND
* UPPER

### CAST{#cast}

The `CAST` function converts an expression to a specified data type. 

#### Syntax

The syntax is as follows:

```sql
CAST ( expression AS data_type )
```

##### expression

`expression` specifies the expression to convert.

> ##### data_type

Reviewer: We still support CAST to FLOAT (and FLOAT is still a valid keyword in OQL). We have only deprecated it as a valid result type for queries, but intermediate calculations should still work.

Reviewer 2: Deprecated or removed entirely?
Is FLOAT deprecated (will it go in Mx 11, for example). I can't really see any point in keeping it if we have DECIMAL.

Reviewer 3: Is there a use case for a new user of OQL to use `FLOAT` instead of `DECIMAL` ?

Reviewer: > Deprecated or removed entirely?

As I said:
> We have only deprecated it as a valid result type for queries, but intermediate calculations should still work.

And indeed, we still have tests for it and it still works, AFAICS.

Either we document here or in the list of keywords, but completely omitting mentioning something that would otherwise break your query if you use as an identifier is not a good idea, IMO. I just wrote the comment here because `CAST` is the only place I remember still accepts `FLOAT`.

Author: I pushed a mention of `FLOAT`

Reviewer: Looks good to me!

Reviewer 3: Wouldn't a better place for the mention of `FLOAT` be in the `CAST` section, given that is the only valid use case?

Author: If we are to move away from Float, I'd not mention any use cases of it at all

Reviewer 3: Sorry, I reread Augusto's comment. Given the reason to mention it is to explain why its a keyword, then the current placement is good. Nothing else needs to be added.

| `DECIMAL`  | Decimal          | 5.3                   | Floating point numeric data                |
| `INTEGER`  | Integer/Long     | 5                     | Integer data                               |
| `LONG`     | Integer/Long     | 5                     | 64 bit width integer data                  |
| `STRING`   | String           | 'my_string'           | Textual data                               |

## Literals

Literals represent values that are constant and are part of the query itself. The supported literals are detailed below:

| Format | Example         | Data Type            | Description                                   |
|--------|-----------------|----------------------|-----------------------------------------------|
|        | `TRUE`, `FALSE` | `BOOLEAN`            | Conditional constants                         |
| 's*'   | 'my_string'     | `STRING`             | String literal                                |
| d+     | 5               | `INTEGER` and `LONG` | Natural number literal                        |
| d+.d+  | 5.3             | `DECIMAL`            | Real number literal                           |

> |        | `NULL`          | N/A                  | NULL literal to represent non-existent values. OQL does not support the keyword `empty`. In XPath, it has identical purpose as `NULL`. |

Reviewer 2: @pijuskri 
Is it a bit confusing to talk about XPath here? Or are you saying that in XPath you can't use `NULL` and have to use `empty` and in OQL you can't use `empty` and have to use `NULL`?

In that case I think it would be better to say:
```suggestion
|        | `NULL`          | N/A                  | NULL literal to represent non-existent values. OQL uses `NULL` where the equivalent XPath expression would use `empty`. |
```

Author: The behavior is that XPath supports both NULL and empty, and OQL supports only NULL, so your proposed change is not entirely correct.

The idea here was to highlight similarities and differences to `XPath Keywords and System Variables` page. We already received some feedback from customers participating in beta testing who are familiar with XPath and who expected some things to work the same way in OQL.

Now that I think of it, system variables have lots of similarities between XPath and OQL and are definitely worth being described from that perspective, but Keywords in XPath are quite different from Literals in OQL, so It's also fine to omit mentioning XPath here.

Reviewer 2: @dahfjkg 
Thanks for getting back to me, Pavel.
I wonder if it is useful to have the comparison somewhere. Would it be worth having an additional page which describes the differences (and similarities) between XPath and OQL? 

I'm happy to put something about XPath here (or get rid of it totally), but if we had it in another place that might be better - and resolve the issues that customers have raised?

Author: The other issues are being addressed by implementing new features, and aside from those and the ones in XPath Keywords and System Variables, not that many points come to mind at the moment. Maybe we can consider that for the future, but I don't think that there is enough reason to work on that now. I'd say let's keep it mentioned here.

Reviewer 2: @dahfjkg - so if we keep it, and my description above is wrong, how about:
```suggestion
|        | `NULL`          | N/A                  | NULL literal to represent non-existent values. OQL only uses `NULL` where the equivalent XPath expression can also use `empty`. |
```
?

## System variables

Most XPath [system variables](/refguide/xpath-keywords-and-system-variables/#system-variables) can be used in OQL with the format:

```sql
'[%SystemVariable%]'
```

> These variables can be used the same way as other expressions.

Reviewer 3: To be more explicit, we should mention that system variables like `%CurrentUser%` are resolved to id's of objects with corresponding type (`LONG` ?). Same with date system variables, are they just strings? How do intervals work (like `%DayLength%`)?

Author: Pushed an update
